File: files/100-199/127/CH-127.py
import pandas as pd
import itertools

path = "CH-127 Add Index Column.xlsx"
input = pd.read_excel(path, usecols="B:C", skiprows=1, nrows=12)
test = pd.read_excel(path, usecols="E:G", skiprows=1, nrows=12).rename(columns=lambda x: x.replace('.1', ''))

input = input.sort_values(by='Stock')
input["index"] = input.groupby(['Stock', (input.groupby('Stock')['Price'].diff() < 0).cumsum()]).cumcount() + 1

print(input.equals(test)) # True

File: files/100-199/127/CH-127.R
library(tidyverse)
library(readxl)

path = "files/CH-127 Add Index Column.xlsx"
input = read_excel(path, range = "B2:C13")
test  = read_excel(path, range = "E2:G13")

compute_index <- function(price_vector) {
  idx <- rep(1, length(price_vector))
  for (i in 2:length(price_vector)) {
    if (price_vector[i] > price_vector[i - 1]) {
      idx[i] <- idx[i - 1] + 1
    } else {
      idx[i] <- 1
    }
  }
  return(idx)
}

result <- input %>%
  group_by(Stock) %>%
  mutate(index = compute_index(Price))

all.equal(result$index, test$index, check.attributes = FALSE)
#> [1] TRUE